### PR TITLE
Support nvidia-docker

### LINKED ISSUE #591: libnvidia-fatbinaryloader.so not found with tensorflow docker image
When running with `tensorflow/tensorflow` docker image from docker hub, I got the following error message

```
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcublas.so locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcudnn.so locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcufft.so locally
I tensorflow/stream_executor/dso_loader.cc:102] Couldn't open CUDA library libcuda.so.1. LD_LIBRARY_PATH: 
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:160] hostname: 6b9e3874a772
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:185] libcuda reported version is: Not found: was unable to find libcuda.so DSO loaded into this program
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:356] driver version file contents: """NVRM version: NVIDIA UNIX x86_64 Kernel Module  367.27  Thu Jun  9 18:53:27 PDT 2016
GCC version:  gcc version 4.8.4 (Ubuntu 4.8.4-2ubuntu1~14.04.3) 
"""
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:189] kernel reported version is: 367.27.0
I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1077] LD_LIBRARY_PATH: 
I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1078] failed to find libcuda.so on this system: Failed precondition: could not dlopen DSO: libcuda.so.1; dlerror: libnvidia-fatbinaryloader.so.367.27: cannot open shared object file: No such file or directory
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcurand.so locally
```

The codalab command is 

```
cl run --request-queue jag --request-docker-image tensorflow/tensorflow:gpu -t "python -c 'import tensorflow; print tensorflow.__version__'"
```

### PR DESCRIPTION
**Pull request by kashizui**
Use nvidia-docker-plugin if available to set up the container creation request for GPU support. Otherwise, fall back to original hacks for GPU support.

Motivation for using nvidia-docker as opposed to other custom fixes:
https://github.com/NVIDIA/nvidia-docker/wiki/Why%20NVIDIA%20Docker

For more information about what nvidia-docker provides:
https://github.com/NVIDIA/nvidia-docker/wiki/Internals

Fixes #591

### CODE REVIEW

File: worker/docker_client.py
Comment (kashizui) [inline worker/docker_client.py]: Place address in string constant

File: worker/docker_client.py
Comment (kashizui) [inline worker/docker_client.py]: yes this is correct

**Comment by kashizui** [comment]
Waiting for feedback on how this works. (thread in #591)

**Comment by percyliang** [comment]
Cool...

1) Per our previous discussion, isn't NVIDIA docker a temporary solution?  I guess the amount of code here isn't too much, so I guess it's fine.

2) Could you explain what nvidia-docker provides over docker?  It seems like we have to do less hacking, which is good.  But in terms of functionality, it seems like we were able to run some GPU jobs with docker, but just not all of them?

**Comment by kashizui** [comment]
Working on this helped me understand what nvidia-docker does even better, and I think I could potentially replicate its functionality. Doing so will allow us to support GPUs on any machine without installing nvidia-docker. 

The problem that nvidia-docker is trying to solve is, allowing Docker images/containers to be agnostic to its hardware re: GPUs. The naive solution would be to install all the NVIDIA drivers in the Docker image. But this will lock the image in with a specific machine configuration (e.g. NVIDIA driver/hardware version). `nvidia-docker` instead finds all the necessary driver files on the host machine, creates hard links to all of them in a Docker volume, and mounts the volume in the container.
https://github.com/NVIDIA/nvidia-docker/wiki/NVIDIA-driver

I'm still stumped as to why our current solution doesn't work, but nvidia-docker does...  I'm trying more experiments to figure  it out. Maybe something involving version mismatches, or needing to mount more than just libcuda? Also not sure why it worked before but not now.

Notes for myself on what the final solution will have to do:

- Ensure that `/dev/nvidia-uvm` is available by calling `nvidia-modprobe -u -c=0`
- Parse the output of `ldconfig -p` to get all the paths of the necessary driver files ([according to logic here](https://github.com/NVIDIA/nvidia-docker/blob/93bb65de7fc349e6de9f27abdaa75875f5572b17/tools/src/nvidia/volumes.go#L173-L211)) and mount them in the container somehow ([example code that creates links in a Docker volume](https://github.com/NVIDIA/nvidia-docker/blob/93bb65de7fc349e6de9f27abdaa75875f5572b17/tools/src/nvidia/volumes.go#L219-L275)). Creating a Docker volume would probably be the best solution (and is more efficient for long-lived workers) since it is probably a bad idea to mount so many volume bindings.

**Comment by kashizui** [comment]
@percyliang do you know what kind of GPU jobs the worker has supported in the past?

**Comment by kashizui** [comment]
Wow the problem was staring me in the eye the entire time (in fact it was the title of the original issue but for some reason I had disregarded it...). It looks like it was because libcuda depends on libnvidia-fatbinaryloader, which was not loaded in our original solution.

```
codalab@jagupard8:~% readelf -d /usr/lib/x86_64-linux-gnu/libcuda.so

Dynamic section at offset 0x71c588 contains 27 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [librt.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libnvidia-fatbinaryloader.so.375.20]
```

File: worker/docker_client.py
Comment (percyliang) [inline worker/docker_client.py:41]: Add some more high-level documentation on trying to use nvidia-docker and what our docker solution is.  Say something about trying to use GPUs.

Separately, on the Wiki, we should add some instructions on how to install/use nvidia-docker.

File: worker/docker_client.py
Comment (percyliang) [inline worker/docker_client.py]: Why the asymmetry (we have a default NVIDIA_DOCKER_HOST but not a DOCKER_HOST)?

**Review by percyliang — APPROVED** [review]
LGTM modulo comments

File: worker/docker_client.py
Comment (kashizui) [reply worker/docker_client.py]: Because the nvidia-docker-plugin server address has a default (https://github.com/NVIDIA/nvidia-docker/wiki/nvidia-docker-plugin) but the docker daemon address doesn't have a default and is often different across hosts

**Pull request merged**
